cmd/ci-secret-generator/main_test.go

```diff
@@ -0,0 +1,391 @@
+						"FieldNum": {
+							"1",
+							"2",
+						},
+						"Env": {
+							"dev",
+							"prod",
+						},
+					},
+				},
+			},
+			expected: []bitWardenItem{
+				{
+					ItemName: "Item1dev",
+					Fields: []fieldGenerator{
+						{
+							Name: "Field1dev",
+							Cmd:  "echo -n Field1dev",
+						},
+					},
+					Attachments: []fieldGenerator{
+						{
+							Name: "Attachment1dev",
+							Cmd:  "echo -n Attachment1dev",
+						},
+					},
+					Attribute: fieldGenerator{
+						Name: "password1dev",
+						Cmd:  "echo -n password1dev",
+					},
+					Notes: "Note1dev",
+					Params: map[string][]string{
+						"FieldNum": {
+							"1",
+							"2",
+						},
+						"Env": {
+							"dev",
+							"prod",
+						},
+					},
+				},
+				{
+					ItemName: "Item1prod",
+					Fields: []fieldGenerator{
+						{
+							Name: "Field1prod",
+							Cmd:  "echo -n Field1prod",
+						},
+					},
+					Attachments: []fieldGenerator{
+						{
+							Name: "Attachment1prod",
+							Cmd:  "echo -n Attachment1prod",
+						},
+					},
+					Attribute: fieldGenerator{
+						Name: "password1prod",
+						Cmd:  "echo -n password1prod",
+					},
+					Notes: "Note1prod",
+					Params: map[string][]string{
+						"FieldNum": {
+							"1",
+							"2",
+						},
+						"Env": {
+							"dev",
+							"prod",
+						},
+					},
+				},
+				{
+					ItemName: "Item2dev",
+					Fields: []fieldGenerator{
+						{
+							Name: "Field2dev",
+							Cmd:  "echo -n Field2dev",
+						},
+					},
+					Attachments: []fieldGenerator{
+						{
+							Name: "Attachment2dev",
+							Cmd:  "echo -n Attachment2dev",
+						},
+					},
+					Attribute: fieldGenerator{
+						Name: "password2dev",
+						Cmd:  "echo -n password2dev",
+					},
+					Notes: "Note2dev",
+					Params: map[string][]string{
+						"FieldNum": {
+							"1",
+							"2",
+						},
+						"Env": {
+							"dev",
+							"prod",
+						},
+					},
+				},
+
+				{
+					ItemName: "Item2prod",
+					Fields: []fieldGenerator{
+						{
+							Name: "Field2prod",
+							Cmd:  "echo -n Field2prod",
+						},
+					},
+					Attachments: []fieldGenerator{
+						{
+							Name: "Attachment2prod",
+							Cmd:  "echo -n Attachment2prod",
+						},
+					},
+					Attribute: fieldGenerator{
+						Name: "password2prod",
+						Cmd:  "echo -n password2prod",
+					},
+					Notes: "Note2prod",
+					Params: map[string][]string{
+						"FieldNum": {
+							"1",
+							"2",
+						},
+						"Env": {
+							"dev",
+							"prod",
+						},
+					},
+				},
+			},
+		},
+	}
+	for _, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+			outputBwItems, err := processBwParameters(tc.inputBwItems)
+			sort.Slice(outputBwItems, func(p, q int) bool {
+				return outputBwItems[p].ItemName < outputBwItems[q].ItemName
+			})
+			equalError(t, tc.expectedError, err)
+			equal(t, tc.expected, outputBwItems)
+		})
+	}
+}
+
+func equalError(t *testing.T, expected, actual error) {
```

Comment:
Why isn't this `reflect.DeepEqual()`?

Reply:
I believe thats to explicitly print out a prettier difference message since we know the types here. This looked like a common pattern in some of the other tools I saw. I can get rid of the equalError and just use the "equal" function below instead.

Reply:
IIRC `cmp.Diff()` will do what you need regardless of type - and the type underneath isn't checked here, either, you just check non-nil `error` interfaces. Unless there is a special, significant reason, this should be 100% of test comparisons:

```go
if diff := cmp.Diff(a, b); diff != "" {
    t.Errorf("%s: got the incorrect thingy: %v", testCase.name, diff)
}
```

Some specific types may require a novel `Comparer` but it's rare.